Unexpected ref safety errors when a discard is used as an `out` argument

**Note after discussion this issue is meant to be tracking the discard case only**

```
using System;

partial class Test
{
    static void Main(){}
    
    void Test1()
    {
        Span<int> a = stackalloc int[10];
        Span<int> b;
        M(ref a, out b);
    }

    void M(ref Span<int> s, out Span<int> s1) => s1 = s;
}
```

Observed:
```
error CS8352: Cannot use local 'a' in this context because it may expose referenced variables outside of their declaration scope
error CS8350: This combination of arguments to 'Test.M(ref Span<int>, out Span<int>)' is disallowed because it may expose variables referenced by parameter 's' outside of their declaration scope
```

Expected: No errors.

The same behavior is around `out` discard:
```
        Span<int> a = stackalloc int[10];
        M(ref a, out _);
``` 

Note that, when local is initialized with stackalloc the errors go away:
```
        Span<int> a = stackalloc int[10];
        Span<int> b = stackalloc int[10];
        M(ref a, out b);
``` 
but there are errors in this case:
```
        Span<int> a = stackalloc int[10];
        Span<int> c = new int[10];
        M(ref a, out c);
```

--- a/src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs
+++ b/src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs
@@ -2167,7 +2167,7 @@ namespace Microsoft.CodeAnalysis.CSharp
                     continue;
                 }
 
-                if (parameter.Type.IsRefLikeType && GetParameterValEscapeLevel(parameter) is { } valEscapeLevel)
+                if (parameter.Type.IsRefLikeType && parameter.RefKind != RefKind.Out && GetParameterValEscapeLevel(parameter) is { } valEscapeLevel)
                 {
                     escapeValues.Add(new EscapeValue(parameter, argument, valEscapeLevel, isRefEscape: false));
                 }
@@ -2417,6 +2417,11 @@ namespace Microsoft.CodeAnalysis.CSharp
 
             void inferDeclarationExpressionValEscape()
             {
+                if (argsOpt.IsDefault)
+                {
+                    return;
+                }
+
                 // find the widest scope that arguments could safely escape to.
                 // use this scope as the inferred STE of declaration expressions.
                 var inferredDestinationValEscape = CallingMethodScope;
@@ -2427,9 +2432,9 @@ namespace Microsoft.CodeAnalysis.CSharp
                         : GetValEscape(fromArg, scopeOfTheContainingExpression));
                 }
 
-                foreach (var (_, fromArg, _, _) in escapeValues)
+                foreach (var argument in argsOpt)
                 {
-                    if (ShouldInferDeclarationExpressionValEscape(fromArg, out var localSymbol))
+                    if (ShouldInferDeclarationExpressionValEscape(argument, out var localSymbol))
                     {
                         localSymbol.SetValEscape(inferredDestinationValEscape);
                     }

--- a/src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs
+++ b/src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs
@@ -2417,11 +2417,6 @@ namespace Microsoft.CodeAnalysis.CSharp
 
             void inferDeclarationExpressionValEscape()
             {
-                if (argsOpt.IsDefault)
-                {
-                    return;
-                }
-
                 // find the widest scope that arguments could safely escape to.
                 // use this scope as the inferred STE of declaration expressions.
                 var inferredDestinationValEscape = CallingMethodScope;

--- a/src/Compilers/CSharp/Test/Semantic/Semantics/RefEscapingTests.cs
+++ b/src/Compilers/CSharp/Test/Semantic/Semantics/RefEscapingTests.cs
@@ -6416,5 +6416,77 @@ public struct Vec4
                 Diagnostic(ErrorCode.ERR_EscapeVariable, "rs6").WithArguments("rs6").WithLocation(45, 16)
                 );
         }
+
+        [Fact, WorkItem(64783, "https://github.com/dotnet/roslyn/issues/64783")]
+        public void OutArgumentsDoNotContributeValEscape_01()
+        {
+            var source = """
+                using System;
+
+                class Program
+                {
+                    static Span<byte> M1()
+                    {
+                        Span<byte> a = stackalloc byte[42];
+                        var ret = OneOutSpanReturnsSpan(out a);
+                        return ret;
+                    }
+
+                    static Span<byte> M2()
+                    {
+                        Span<byte> a = stackalloc byte[42];
+                        TwoOutSpans(out a, out Span<byte> b);
+                        return b;
+                    }
+
+                    static Span<byte> OneOutSpanReturnsSpan(out Span<byte> a)
+                    {
+                        // 'return a' is illegal until it is overwritten
+                        a = default;
+                        return default;
+                    }
+
+                    static void TwoOutSpans(out Span<byte> a, out Span<byte> b)
+                    {
+                        // 'a = b' and 'b = a' are illegal until one has already been written
+                        a = b = default;
+                    }
+                }
+
+                """;
+
+            var comp = CreateCompilationWithSpan(source);
+            comp.VerifyDiagnostics();
+        }
+
+        [Fact, WorkItem(56587, "https://github.com/dotnet/roslyn/issues/56587")]
+        public void OutArgumentsDoNotContributeValEscape_02()
+        {
+            // Test that out discard arguments are not treated as inputs.
+            // This means we don't need to take special care to zero-out the variable used for a discard argument between uses.
+            var source = """
+                using System;
+
+                class Program
+                {
+                    static Span<byte> M1()
+                    {
+                        Span<byte> a = stackalloc byte[42];
+                        TwoOutSpans(out a, out _);
+                        TwoOutSpans(out _, out Span<byte> c);
+                        return c;
+                    }
+
+                    static void TwoOutSpans(out Span<byte> a, out Span<byte> b)
+                    {
+                        // 'a = b' and 'b = a' are illegal until one has already been written
+                        a = b = default;
+                    }
+                }
+                """;
+
+            var comp = CreateCompilationWithSpan(source);
+            comp.VerifyDiagnostics();
+        }
     }
 }

--- a/src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs
+++ b/src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs
@@ -14229,18 +14229,9 @@ class Enumerator2
 
             comp = CreateCompilation(source);
             comp.VerifyEmitDiagnostics(
-                // (7,18): error CS8352: Cannot use variable '(scoped R r1, scoped var _)' in this context because it may expose referenced variables outside of their declaration scope
-                //         foreach ((scoped R r1, scoped var _) in new Enumerable1()) break;
-                Diagnostic(ErrorCode.ERR_EscapeVariable, "(scoped R r1, scoped var _)").WithArguments("(scoped R r1, scoped var _)").WithLocation(7, 18),
                 // (7,32): error CS9061: The 'scoped' modifier cannot be used with discard.
                 //         foreach ((scoped R r1, scoped var _) in new Enumerable1()) break;
                 Diagnostic(ErrorCode.ERR_ScopedDiscard, "scoped").WithLocation(7, 32),
-                // (7,49): error CS8350: This combination of arguments to 'R.Deconstruct(out R, out R)' is disallowed because it may expose variables referenced by parameter 'x' outside of their declaration scope
-                //         foreach ((scoped R r1, scoped var _) in new Enumerable1()) break;
-                Diagnostic(ErrorCode.ERR_CallArgMixing, "new Enumerable1()").WithArguments("R.Deconstruct(out R, out R)", "x").WithLocation(7, 49),
-                // (8,18): error CS8352: Cannot use variable '(scoped ref R r2, scoped ref var _)' in this context because it may expose referenced variables outside of their declaration scope
-                //         foreach ((scoped ref R r2, scoped ref var _) in new Enumerable2(ref r)) break;
-                Diagnostic(ErrorCode.ERR_EscapeVariable, "(scoped ref R r2, scoped ref var _)").WithArguments("(scoped ref R r2, scoped ref var _)").WithLocation(8, 18),
                 // (8,26): error CS9072: A deconstruction variable cannot be declared as a ref local
                 //         foreach ((scoped ref R r2, scoped ref var _) in new Enumerable2(ref r)) break;
                 Diagnostic(ErrorCode.ERR_DeconstructVariableCannotBeByRef, "ref").WithLocation(8, 26),
@@ -14250,12 +14241,6 @@ class Enumerator2
                 // (8,43): error CS9072: A deconstruction variable cannot be declared as a ref local
                 //         foreach ((scoped ref R r2, scoped ref var _) in new Enumerable2(ref r)) break;
                 Diagnostic(ErrorCode.ERR_DeconstructVariableCannotBeByRef, "ref").WithLocation(8, 43),
-                // (8,57): error CS8350: This combination of arguments to 'R.Deconstruct(out R, out R)' is disallowed because it may expose variables referenced by parameter 'x' outside of their declaration scope
-                //         foreach ((scoped ref R r2, scoped ref var _) in new Enumerable2(ref r)) break;
-                Diagnostic(ErrorCode.ERR_CallArgMixing, "new Enumerable2(ref r)").WithArguments("R.Deconstruct(out R, out R)", "x").WithLocation(8, 57),
-                // (9,18): error CS8352: Cannot use variable '(scoped ref readonly R r5, scoped ref readonly var _)' in this context because it may expose referenced variables outside of their declaration scope
-                //         foreach ((scoped ref readonly R r5, scoped ref readonly var _) in new Enumerable2(ref r)) break;
-                Diagnostic(ErrorCode.ERR_EscapeVariable, "(scoped ref readonly R r5, scoped ref readonly var _)").WithArguments("(scoped ref readonly R r5, scoped ref readonly var _)").WithLocation(9, 18),
                 // (9,26): error CS9072: A deconstruction variable cannot be declared as a ref local
                 //         foreach ((scoped ref readonly R r5, scoped ref readonly var _) in new Enumerable2(ref r)) break;
                 Diagnostic(ErrorCode.ERR_DeconstructVariableCannotBeByRef, "ref").WithLocation(9, 26),
@@ -14265,9 +14250,6 @@ class Enumerator2
                 // (9,52): error CS9072: A deconstruction variable cannot be declared as a ref local
                 //         foreach ((scoped ref readonly R r5, scoped ref readonly var _) in new Enumerable2(ref r)) break;
                 Diagnostic(ErrorCode.ERR_DeconstructVariableCannotBeByRef, "ref").WithLocation(9, 52),
-                // (9,75): error CS8350: This combination of arguments to 'R.Deconstruct(out R, out R)' is disallowed because it may expose variables referenced by parameter 'x' outside of their declaration scope
-                //         foreach ((scoped ref readonly R r5, scoped ref readonly var _) in new Enumerable2(ref r)) break;
-                Diagnostic(ErrorCode.ERR_CallArgMixing, "new Enumerable2(ref r)").WithArguments("R.Deconstruct(out R, out R)", "x").WithLocation(9, 75),
                 // (11,35): error CS9061: The 'scoped' modifier cannot be used with discard.
                 //         foreach ((scoped var r11, scoped R _) in new Enumerable1()) break;
                 Diagnostic(ErrorCode.ERR_ScopedDiscard, "scoped").WithLocation(11, 35),

